src/codegen.cpp

```diff
@@ -4376,6 +4385,25 @@ static void emit_varinfo_assign(jl_codectx_t &ctx, jl_varinfo_t &vi, jl_cgval_t
     }
 }
 
+static void emit_binding_store(jl_codectx_t &ctx, jl_binding_t *bnd, Value *bp, jl_value_t *r, ssize_t ssaval, AtomicOrdering Order)
+{
+    assert(bnd);
+    jl_cgval_t rval_info = emit_expr(ctx, r, ssaval);
+    Value *rval = boxed(ctx, rval_info);
+    if (!bnd->constp && bnd->ty && jl_subtype(rval_info.typ, bnd->ty)) {
+        StoreInst *v = ctx.builder.CreateAlignedStore(rval, bp, Align(sizeof(void*)));
+        v->setOrdering(Order);
+        tbaa_decorate(ctx.tbaa().tbaa_binding, v);
+        ctx.builder.CreateAssumption(ctx.builder.CreateIsNotNull(rval));
```

Comment:
I was hoping this would elide the undef var check in the following example, but that doesn't seem to work:
```julia
julia> global y::Int

julia> function g()
           global y = 1
           y
       end
g (generic function with 1 method)

julia> @code_llvm g()
;  @ REPL[2]:1 within `g`
define i64 @julia_g_1078() #0 {
top:
;  @ REPL[2]:2 within `g`
  store atomic {}* inttoptr (i64 140699562991712 to {}*), {}** inttoptr (i64 140696999452504 to {}**) unordered, align 8
  call void @llvm.assume(i1 icmp ne ({}* inttoptr (i64 140699562991712 to {}*), {}* null))
;  @ REPL[2]:3 within `g`
  br i1 icmp ne ({}* inttoptr (i64 140699562991712 to {}*), {}* null), label %ok, label %err

err:                                              ; preds = %top
  call void @ijl_undefined_var_error({}* inttoptr (i64 140699564102024 to {}*))
  unreachable

ok:                                               ; preds = %top
  %0 = load i64, i64* inttoptr (i64 140699562991712 to i64*), align 8
  ret i64 %0
}
```

Reply:
Of course, the `llvm.assume` doesn't make much sense here, since it's a constant pointer so the null check should be trivial. Could this be a matter of us not running the right optimization passes? For reference, I can eliminate branches like this in godbolt by compiling with `-O1`: https://godbolt.org/z/sE5hs13xe

Reply:
Hmm, that doesn't seem to contain anything too interesting in that regard:
```
julia> @code_llvm raw=true dump_module=true g()
; ModuleID = 'g'
source_filename = "g"
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

;  @ REPL[2]:1 within `g`
define i64 @julia_g_1267() #0 !dbg !5 {
top:
;  @ REPL[2]:2 within `g`
  store atomic {} addrspace(10)* addrspacecast ({}* inttoptr (i64 139739127959648 to {}*) to {} addrspace(10)*), {} addrspace(10)** inttoptr (i64 139739075273112 to {} addrspace(10)**) unordered, align 8, !dbg !7, !tbaa !8
;  @ REPL[2]:3 within `g`
  br i1 icmp ne ({} addrspace(10)* addrspacecast ({}* inttoptr (i64 139739127959648 to {}*) to {} addrspace(10)*), {} addrspace(10)* null), label %ok, label %err, !dbg !13

err:                                              ; preds = %top
  call void @ijl_undefined_var_error({} addrspace(12)* addrspacecast ({}* inttoptr (i64 139739129125664 to {}*) to {} addrspace(12)*)), !dbg !13
  unreachable, !dbg !13

ok:                                               ; preds = %top
  %0 = load i64, i64 addrspace(10)* addrspacecast (i64* inttoptr (i64 139739127959648 to i64*) to i64 addrspace(10)*), align 8, !dbg !13, !tbaa !14
  ret i64 %0, !dbg !13
}

define nonnull {} addrspace(10)* @jfptr_g_1268({} addrspace(10)* %0, {} addrspace(10)** %1, i32 %2) #1 {
top:
  %3 = call i64 @julia_g_1267() #0
  %4 = call nonnull {} addrspace(10)* @ijl_box_int64(i64 signext %3)
  ret {} addrspace(10)* %4
}

declare nonnull {} addrspace(10)* @ijl_box_int64(i64 signext)

; Function Attrs: inaccessiblememonly norecurse nounwind
declare void @julia.write_barrier_binding({} addrspace(10)* readonly, ...) #2

; Function Attrs: noreturn
declare void @ijl_undefined_var_error({} addrspace(12)*) #3

; Function Attrs: inaccessiblemem_or_argmemonly
declare void @ijl_gc_queue_root({} addrspace(10)*) #4

; Function Attrs: inaccessiblemem_or_argmemonly
declare void @jl_gc_queue_binding({} addrspace(10)*) #4

; Function Attrs: allocsize(1)
declare noalias nonnull {} addrspace(10)* @ijl_gc_pool_alloc(i8*, i32, i32) #5

; Function Attrs: allocsize(1)
declare noalias nonnull {} addrspace(10)* @ijl_gc_big_alloc(i8*, i64) #5

attributes #0 = { "probe-stack"="inline-asm" }
attributes #1 = { "probe-stack"="inline-asm" "thunk" }
attributes #2 = { inaccessiblememonly norecurse nounwind }
attributes #3 = { noreturn }
attributes #4 = { inaccessiblemem_or_argmemonly }
attributes #5 = { allocsize(1) }

!llvm.module.flags = !{!0, !1}
!llvm.dbg.cu = !{!2}

!0 = !{i32 2, !"Dwarf Version", i32 4}
!1 = !{i32 2, !"Debug Info Version", i32 3}
!2 = distinct !DICompileUnit(language: DW_LANG_Julia, file: !3, producer: "julia", isOptimized: true, runtimeVersion: 0, emissionKind: NoDebug, enums: !4, nameTableKind: GNU)
!3 = !DIFile(filename: "REPL[2]", directory: ".")
!4 = !{}
!5 = distinct !DISubprogram(name: "g", linkageName: "julia_g_1267", scope: null, file: !3, line: 1, type: !6, scopeLine: 1, spFlags: DISPFlagDefinition | DISPFlagOptimized, unit: !2, retainedNodes: !4)
!6 = !DISubroutineType(types: !4)
!7 = !DILocation(line: 2, scope: !5)
!8 = !{!9, !9, i64 0}
!9 = !{!"jtbaa_binding", !10, i64 0}
!10 = !{!"jtbaa_data", !11, i64 0}
!11 = !{!"jtbaa", !12, i64 0}
!12 = !{!"jtbaa"}
!13 = !DILocation(line: 3, scope: !5)
!14 = !{!15, !15, i64 0}
!15 = !{!"jtbaa_value", !10, i64 0}
```

Or do I need a special build to see this in more detail?